os/arch/arm/src/amebad/amebad_serial.c

```diff
@@ -571,6 +573,22 @@ static int rtl8721d_up_ioctl(FAR struct uart_dev_s *dev, int cmd, unsigned long
 		serial_control_loopback(sdrv[uart_index_get(priv->tx)], *(bool *)arg);
 		break;
 
+	case TIOCS_AVAIL:
+		g_tc_rxavailable = 1;
+		g_tc_rxavailable_res = 0;
+		rtl8721d_up_send(dev, 'a');
+		sleep(1);
+		ret = g_tc_rxavailable_res;
```

Comment:
I don't know what you want to do. Could you leave comment to explain?

Reply:
for rxavailable TC, steps are:

- rxavailable returns true when there is data in rx fifo and when data comes on the tx line in loopback mode then it enables the RxIrq and it goes to uart_recvchars(), where it checks whether data is available in rx fifo or not.
- so to check the data in rx fifo, we need to send the data first, which we have sent through rtl8721d_up_send() and before sending the data you need to enable the g_tc_rxavailable so that g_tc_rxavailable_res can collect the result for our sent data and when it collects the result then immediately turn off the g_tc_rxavailable, which means it won't collect the result further.

so, this algorithm came to my mind to test rxavailable(), if you have other steps, which can be better, please suggest.
I'll make comment once we agree on the algo.

Reply:
@kishore-sn @sangam-swami @jeongchanKim please review.

Reply:
1. I understand they are "begin" and "end" of tc to get valid data. Am I right?
2. It's the loopback test. There is no way to get another data. Why do you need begin and end?
3. Even it's loopback mode, sender and receiver should be implemented using existed APIs without more variables. If it's impossible, it means our APIs are not enough to use serial functionality. Then we should add them as public API, not tc.

Reply:
How about changing this implementation as follows:
Remove all global variables. 

`case TIOCS_AVAIL:`
`rtl8721d_up_send(dev, 'a');`
	`	sleep(1);`
	`	ret = rtl8721d_up_rxavailable(dev);`
`break;`

Reply:
On the other hand, I think we can change it further to make the code much cleaner.

In the driver we do:
`case TIOCS_AVAIL:`
`ret = rtl8721d_up_rxavailable(dev);`
`break;`

In the test case file we do:
`iotbus_uart_context_h h_uart = iotbus_uart_init(DEVPATH);`
`	ret = iotbus_uart_control_loopback(h_uart, LOOPBACK_ON);`
`		ret = iotbus_uart_write(h_uart, sz_input_text, sizeof(sz_input_text));`
`ret = iotbus_uart_rxavailable(h_uart);`

Reply:
we can not do like this:
```
case TIOCS_AVAIL:
rtl8721d_up_send(dev, 'a');
sleep(1);
ret = rtl8721d_up_rxavailable(dev);
break;
```

because when we sent the data in loopback mode then it comes on the RX FIFO immediately and generates the interrupt,
and calls uart_recvchars() where it checks whether data is available in RX FIFO or not, if data is available then it removes the data from RX FIFO and copies it to buffer so the following lines won't work, because when it comes to rtl8721d_up_rxavailable(), then data is not in RX FIFO because it has been transferred to buffer.
```
ret = rtl8721d_up_rxavailable(dev);
break;
```

Reply:
ok. 
Then we need to keep a global variable. But its better to keep a single global variable and protect it with a semaphore. Add a new ioctl to reset global variable. 

`bool g_rxavail'
'sem_t g_rxavail_sem`

`TCIOC_RXAVAIL_RESET:`
`take g_rxavail_sem`
`g_rxavail = false`
`give g_rxavail_sem`

`TCIOC_RXAVAIL:`
`take g_rxavail_sem`
`ret = g_rxavail;`
`give g_rxavail_sem`

And in rtl8721d_up_rxavailable()

`take g_rxavail_sem`
`g_rxavail = ret;`
`give g_rxavail_sem`

Reply:
ok. 
Then we need to keep a global variable. But its better to keep a single global variable and protect it with a semaphore. Add a new ioctl to reset global variable. 

`bool g_rxavail'
'sem_t g_rxavail_sem`

`TCIOC_RXAVAIL_RESET:`
`take g_rxavail_sem`
`g_rxavail = false`
`give g_rxavail_sem`

`TCIOC_RXAVAIL:`
`take g_rxavail_sem`
`ret = g_rxavail;`
`give g_rxavail_sem`

And in rtl8721d_up_rxavailable()

`take g_rxavail_sem`
`g_rxavail = ret;`
`give g_rxavail_sem`